react/index.js

```diff
@@ -1,27 +1,155 @@
+      autoName: '',
+      autoAuthors: '',
+      cachedID: '',
+      cachedName: '',
+      cachedAuthors: '',
+    }
+
+    this.state = Object.assign({}, this.clearedState)
+  }
+
+  deleteBook(bookType) {
+    const {deleteAutopersistedBook, deleteCachedBook} = this.props
+    let mutate, id
+
+    if (bookType === AUTOPERSISTED) {
+      mutate = deleteAutopersistedBook
+      id = this.state.autoID
+    } else {
+      mutate = deleteCachedBook
+      id = this.state.cachedID
+    }
+
+    const deleteData = {
+      variables: {
+        id: id
+      },
+      refetchQueries: [{query: getBooks}]
+    }
+
+    mutate(deleteData).then(this.clearInput)
+  }
+
+  addNewBook(bookType) {
+    const {createAutopersistedBook, createCachedBook} = this.props
+    let mutate, name, authors
+
+    if (bookType === AUTOPERSISTED) {
+      mutate = createAutopersistedBook
+      name = this.state.autoName
+      authors = this.state.autoAuthors
+    } else {
+      mutate = createCachedBook
+      name = this.state.cachedName
+      authors = this.state.cachedAuthors
+    }
+
+    const addData = {
+      variables: {
+        name: name,
+        authors: authors.split(',')
+      },
+      refetchQueries: [{query: getBooks}]
+    }
+
+    mutate(addData).then(this.clearInput)
+  }
+
+  clearInput() {
+    this.setState(this.clearedState)
+  }
+
+  render() {
+    const {data: {autopersistedBooks, cachedBooks}} = this.props
+    const inlineStyle = {
+      padding: '15px'
+    }
+
+    return (
+      <div style={inlineStyle}>
+        <h2>Autopersisted Books</h2>
+        <h3>Books in the Library</h3>
+        {autopersistedBooks && autopersistedBooks.map(({id, name, authors}) => {
+          return <div key={id}>
+                    <b>{name || 'Name me, please'}</b>
+                    <br/>Authors: {(authors && authors.map((authors) => `${authors}, `)) || 'Give me a father, please'}
+                    <br/>ID (click to select):<div onClick={event => this.setState({autoID: id})}>{id || 'I should have an ID, this is embarrassing'}</div>
+                    <br/>
+                </div>
+        })}
+        <br/><hr/>
+
+        <h3>Add a book to the library</h3>
+        <label>Name: <input type="text" value={this.state.autoName} onChange={event => this.setState({autoName: event.target.value})} /></label>
+        <label>Authors(split by ','): <input type="test" value={this.state.autoAuthors} onChange={event => this.setState({autoAuthors: event.target.value})} /></label>
+        <button onClick={event => this.addNewBook(AUTOPERSISTED)}>Insert</button>
+        <br/><hr/>
+
+        <h3>Remove a book from the library</h3>
+        <label>ID:<input type="text" value={this.state.autoID} onChange={event => this.setState({autoID: event.target.value})}/></label>
+        <button onClick={event => this.deleteBook(AUTOPERSISTED)}>DELETE</button>
+        <br/><hr/>
+
+
+        <h2>Cached Books</h2>
+        <h3>Books in the Library</h3>
```

Comment:
This block is almost the same as the previous one, so it'd be nice to refactor into a separate `Component`, making this example much easier to read (and maintain). It could receive everything specific it would need to handle different types of books. So instead of all this, this `render` function could be just something like:

```js
<div style={inlineStyle}>
  <BooksManager
    title="Auto Persisted Books"
    books={autoPersistedBooks}
    onCreate={createAutoPersistedBook}
    onDelete={deleteAutoPersistedBook}
  />
  <BooksManager
    title="Cached Books"
    books={cachedBooks}
    onCreate={createCachedBook}
    onDelete={deleteCachedBook}
  />
</div>
```

With this, there will also be no need to track different variables for auto persisted vs cached input data in the component's state. `BooksManager` can track its own state for each type, making the code cleaner again.

That's one of the advantages of React, being able to easily modularize your code into separate components 🎉!